File: leetcode/cpp/57.cpp
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> insert(std::vector<std::vector<int>>& intervals, std::vector<int>& newInterval) {
        std::vector<std::vector<int>> result;
        int i = 0;
        int n = intervals.size();

        // Case 1: Add intervals that end before the new interval starts
        while (i < n && intervals[i][1] < newInterval[0]) {
            result.push_back(intervals[i]);
            i++;
        }

        // Case 2: Merge overlapping intervals
        while (i < n && intervals[i][0] <= newInterval[1]) {
            newInterval[0] = std::min(newInterval[0], intervals[i][0]);
            newInterval[1] = std::max(newInterval[1], intervals[i][1]);
            i++;
        }
        result.push_back(newInterval);

        // Case 3: Add remaining intervals
        while (i < n) {
            result.push_back(intervals[i]);
            i++;
        }

        return result;
    }
};

File: leetcode/go/57.go
package main

func insert(intervals [][]int, newInterval []int) [][]int {
	var ans [][]int
	n := len(intervals)
	for i := 0; i < n; i++ {
		if intervals[i][1] < newInterval[0] ||
			intervals[i][0] > newInterval[1] {
			//interval comes before the newInterval,
			//2.Interval comes wholly after the newInterval
			ans = append(ans, intervals[i])
		} else {
			//we have to merging
			start := intervals[i][0]
			end := newInterval[1]
			for intervals[i][0] <= end {
				end = max(intervals[i][1], end)
				i++
			}
			ans = append(ans, []int{start, end})
		}
	}
	return ans
}
